### PR TITLE
IGNITE-22808 Migrate running node missed during CMG reparation to new cluster

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-22808

### CODE REVIEW

File: modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java [hunk 1]
```diff
@@ -111,6 +128,11 @@ private CompletableFuture<Void> restartOrShutdownFuture(int nodeIndex) {
         return ((IgniteServerImpl) cluster.server(nodeIndex)).restartOrShutdownFuture();
     }
 
+    private static ClusterState clusterState(IgniteImpl restartedIgniteImpl1)
+            throws InterruptedException, ExecutionException, TimeoutException {
+        return restartedIgniteImpl1.clusterManagementGroupManager().clusterState().get(10, SECONDS);
```
Comment (tkalkirill) [inline modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java:133]: Why not the classic `assertThat(fut, willCompleteSuccessfully())` and then `fut.join()`?

File: modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java [hunk 2]
```diff
@@ -223,4 +245,110 @@ void nodesThatSawNoReparationHaveSeparatePhysicalTopology() throws Exception {
...
+        cluster.stopNode(0);
+
+        // Repair CMG with just node 1.
+        initiateCmgRepairVia(igniteImpl(1), 1);
+        IgniteImpl restartedIgniteImpl1 = waitTillNodeRestartsInternally(1);
+        waitTillCmgHasMajority(restartedIgniteImpl1);
+
+        migrate(0, 1);
+
+        LogicalTopologySnapshot topologySnapshot = igniteImpl(1).logicalTopologyService().logicalTopologyOnLeader().get(10, SECONDS);
```
Comment (tkalkirill) [inline modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java:267]: Same about `assertThat(fut, ...)`

File: modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java [hunk 3]
```diff
@@ -223,4 +245,110 @@ void nodesThatSawNoReparationHaveSeparatePhysicalTopology() throws Exception {
...
+        // Stop the majority of CMG.
+        IntStream.of(0, 1).parallel().forEach(cluster::stopNode);
+
+        // Repair CMG with nodes 2, 3, 4.
+        initiateCmgRepairVia(igniteImpl(2), 2, 3, 4);
+        IgniteImpl restartedIgniteImpl2 = waitTillNodeRestartsInternally(2);
+        waitTillCmgHasMajority(restartedIgniteImpl2);
+
+        // Starting the nodes that did not see the repair.
+        List<IgniteImpl> partialNodes = IntStream.of(0, 1).parallel()
```
Comment (tkalkirill) [inline modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java:310]: why `.parallel()` ?

File: modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java [hunk 4]
```diff
@@ -223,4 +245,110 @@ void nodesThatSawNoReparationHaveSeparatePhysicalTopology() throws Exception {
...
+                .mapToObj(index -> ((IgniteServerImpl) cluster.startEmbeddedNode(index).server()).igniteImpl())
+                .collect(toList());
+
+        initiateMigrationToNewCluster(partialNodes.get(0), igniteImpl(2));
+
+        waitTillNodeRestartsInternally(0);
+        waitTillNodeRestartsInternally(1);
+
+        LogicalTopologySnapshot topologySnapshot = igniteImpl(2).logicalTopologyService().logicalTopologyOnLeader().get(10, SECONDS);
```
Comment (tkalkirill) [inline modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java:319]: Same about `assertThat(fut, ...)`

File: modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java (hunk 1 above)
Comment (rpuch) [reply modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java:133]: What you suggest requires 3 lines:
```
CompletableFuture<T> future = blablaFuture();
assertThat(future, willCompleteSuccessfully());
return future.join();
```
But with timed get we only need one line:
```
return blablaFuture().get(...);
```
The matcher-based approach is good when you don't need the result of the future, but here produces a code that is too cumbersome.

File: modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java (hunk 3 above)
Comment (rpuch) [reply modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java:310]: To start them in parallel and save some test time

File: modules/system-disaster-recovery/src/main/java/org/apache/ignite/internal/disaster/system/SystemDisasterRecoveryManagerImpl.java [hunk 5]
```diff
@@ -255,6 +257,37 @@ private static boolean isMajorityOfCmgAreSuccesses(
...
+                    rethrowIfError(ex);
+
+                    restarter.initiateRestart();
+                    return null;
+                }, restartExecutor);
+    }
+
+    private ResetClusterMessage buildResetClusterMessageForMigrate(ClusterState clusterState) {
+        List<UUID> formerClusterIds = clusterState.formerClusterIds();
+        assert formerClusterIds != null;
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/system-disaster-recovery/src/main/java/org/apache/ignite/internal/disaster/system/SystemDisasterRecoveryManagerImpl.java]: Added validation and improved assertion message

File: modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java (hunk 3 above)
Comment (tkalkirill) [reply modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java:310]: Please leave comment about it.

File: modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java (hunk 1 above)
Comment (tkalkirill) [reply modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java:133]: We use this construction quite often, I agree that it requires more lines of code. I don't insist, it's up to you.
Example: `org.apache.ignite.internal.cluster.management.ItClusterManagerTest#findLeaderNode`

**Review by tkalkirill — APPROVED** [review]

File: modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java (hunk 1 above)
Comment (rpuch) [reply modules/system-disaster-recovery/src/integrationTest/java/org/apache/ignite/internal/disaster/system/ItCmgDisasterRecoveryTest.java:133]: `.get(...)` declares checked exceptions. In `findLeaderNode()` we would have to handle them with a try/catch, bloating the 'just one line' to 'many lines'. But I used this in contexts where we can just declare those `throws` ourselves, so the bloating does not happen.

**Pull request merged**